Core/Parser/SchemaParser.cs

```diff
@@ -207,7 +237,42 @@ public async Task<ISchema> Parse()
                 {
                     throw new UnrecognizedTypeException(typeToken, definitionToken.Lexeme);
                 }
+                var reference = _definitions[typeToken.Lexeme];
+                var referenceScope = reference.Scope;
+                var definition = _definitions[definitionToken.Lexeme];
+                var definitionScope = definition.Scope;
+                // You're not allowed to reference types declared within a union from elsewhere
+                // Check if reference has a union in scope but definition does not have the same union in scope
+                // Throw ReferenceScopeException if so
+
+                // It might be better for this to go inside BebopSchema.Validate but it's much simpler if I can use the typeReferences map
+                if (referenceScope is not null && referenceScope.Find((parent) => parent is UnionDefinition) is UnionDefinition union)
+                {
+                    if (definitionScope is null || !definitionScope.Contains(union))
+                    {
```

Comment:
You can get rid of these `null` checks after applying the change in my earlier comment about `Scope`.